src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+                {
+                    this.UpdatedReplaceableResolutions(mayBeReplaced, combinedBuilder, typeParam, paramRes);
+                }
+
+                // Validate and add each resolution to the result.
+                foreach (var (typeParam, paramRes) in resolvedDictionary)
+                {
+                    // Check that we are not constricting a type parameter to another type parameter of the same callable.
+                    this.UpdateConstrictionFlag(typeParam, paramRes);
+
+                    // Check that there is no conflicting resolution already defined.
+                    if (!this.SummarizesOverConflictingResolution)
+                    {
+                        this.SummarizesOverConflictingResolution = combinedBuilder.TryGetValue(typeParam, out var current)
+                            && !current.Equals(paramRes) && !IsSelfResolution(typeParam, current);
+                    }
+
+                    // Add the resolution to the current dictionary.
+                    combinedBuilder[typeParam] = paramRes;
+                }
+            }
+
+            this.CombinedTypeParameterResolutions = combinedBuilder.ToImmutable();
+        }
+
+        // Nested Classes
+
+        /// <summary>
+        /// Walker that collects all of the type parameter references for a given ResolvedType
+        /// and returns them as a HashSet.
+        /// </summary>
+        private class GetTypeParameters : TypeTransformation<GetTypeParameters.TransformationState>
+        {
+            /// <summary>
+            /// Walks the given ResolvedType and returns all of the type parameters referenced.
+            /// </summary>
+            public static HashSet<TypeParameterName> Apply(ResolvedType res)
+            {
+                var walker = new GetTypeParameters();
+                walker.OnType(res);
+                return walker.SharedState.TypeParams;
+            }
+
+            internal class TransformationState
+            {
+                public HashSet<TypeParameterName> TypeParams = new HashSet<TypeParameterName>();
+            }
+
+            private GetTypeParameters() : base(new TransformationState(), TransformationOptions.NoRebuild)
+            {
+            }
+
+            private static TypeParameterName AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+
+            public override ResolvedTypeKind OnTypeParameter(QsTypeParameter tp)
+            {
+                this.SharedState.TypeParams.Add(AsTypeResolutionKey(tp));
+                return base.OnTypeParameter(tp);
+            }
+        }
+
+        /// <summary>
+        /// Walker that checks a given type parameter resolution to see if it constricts
+        /// the type parameter to another type parameter of the same callable.
+        /// </summary>
+        private class ConstrictionCheck : TypeTransformation<ConstrictionCheck.TransformationState>
```

Comment:
Similar feedback regarding the name of this class. Maybe `ConstrictionChecker` would better signal that it is a class.

Reply:
Since transformations are supposed to be verbs, I should actually probably rename this to something like `CheckForConstriction`.